### ISSUE TITLE
wrap more than just the command (e.g `git checkout`)

### ISSUE THREAD

**Issue by despairblue** [fish-shell/fish-shell#1976]
Is this possible?:

```
function gco --wraps 'git checkout'
  git checkout $argv
end
```

I would expect this to auto complete like git checkout, though it does not.

**Comment by ridiculousfish** [comment]
Not yet. That looks for a command `git\ checkout`, i.e. with an embedded space.

Today commands can only wrap other commands. The reason is that I wasn't sure how far to take it - can you also wrap commands with subcommands, redirections, unclosed quotes, etc? Taken to extremes you wind up with something like zsh's no-holds-barred global aliases.

I think your use case is very reasonable. If you want to think through the issues and come up with a proposal for how it should behave, I'd sure appreciate it.

You might also try an abbreviation here.

**Comment by despairblue** [comment]
yeah I found out about the abbreviations in a related issue. I tried them
and they suit this use case actually pretty well. Is there any ETA when
they will be released?

**Comment by zanchey** [comment]
Not yet.

**Comment by simnalamburt** [comment]
**+1** for this issue. I'm former zsh user and trying to migrate to the fish shell, but blocked by this. I lived with [tons of git aliases](https://github.com/simnalamburt/pkg-cgitc/blob/master/init.fish#L41), and now I have serious problem with using them. I desperately need this feature.

```fish
alias g='git'

alias ga='git add'
alias gaa='git add --all'
alias gapa='git add --patch'

alias gb='git branch'
alias gba='git branch -a'
alias gbl='git blame -b -w'
alias gbnm='git branch --no-merged'
alias gbr='git branch --remote'
alias gbs='git bisect'
...
```

**Comment by pickfire** [comment]
@simnalamburt You can use `abbr` too, it will automatically lengthen it.
```fish
abbr -a g git
```

**Comment by krader1961** [comment]
@simnalamburt for aliases meant only for interactive use you're probably better served by the `abbr` command.

**Comment by simnalamburt** [comment]
I used to make about 200 aliases. I tried to change all those aliases into `abbr` but it made fish too slow.

https://github.com/simnalamburt/pkg-cgitc/blob/master/init.fish

**Comment by faho** [comment]
@simnalamburt: Too slow to start or slow while running?

**Comment by simnalamburt** [comment]
Too slow to start

**Comment by faho** [comment]
In that case, I've noticed abbr.fish taking a while. I'll try to take a look.

**Comment by pickfire** [comment]
@faho, nice. Try to open it in a new issue.

**Comment by faho** [comment]
@simnalamburt: You may want to try one of two things:

- Set the abbreviations once and remove the `abbr` calls from your config file

This is because abbreviations are stored in a universal (by default) variable shared among fish instances, so you don't need to set them up each time fish starts. You could also keep the calls in your config file but guard them with

```fish
if set -q fish_initialized
   set -U fish_initialized
   abbr so wow
   #...
end
```

- #2654 

This is only a speedup of about 1/3 in the common case (abbrs already written) and depends on a git master fish (because of the `string` builtin), but maybe it's enough for you?

**Comment by simnalamburt** [comment]
@faho Thanks. Greatly settled with below:
```fish
if not set -q cgitc_initialized
  set -U cgitc_initialized

  # Abbrs
end
```

**Comment by zanchey** [comment]
I think abbreviations are the right answer here - I don't know that there's much more to do. Thoughts?

**Comment by faho** [comment]
@zanchey: I don't think so. That wouldn't work for commands that wrap other commands, for one, and sometimes you have a function that can take the exact same arguments as something else, but you don't want it to show up fully in the commandline.

**Comment by faho** [comment]
For example, I have this function:

```fish
function duthis
	set -q argv[1]; or set argv * .*
        du -sch $argv | sort -h
end
```

Of course since the arguments are passed to `du -sch`, it should be completed like `du -sch`. But I don't want that thing as an abbr instead (and currently it wouldn't actually work as an abbr since it would put the cursor after the `sort` call).

(Currently, wrapping `du -sch` would behave the same as wrapping `du`, but that's because of #1315).

>An example of what I mean: what should cmd --wraps 'foo > bar' do?

@ridiculousfish: This is only about completions, right? Those aren't affected by redirections. Basically what I think would solve this completely is to use the argument to "--wraps" as the fake commandline and pass that to the completions, just like if `complete -C` were called.

**Comment by ridiculousfish** [comment]
The tricky part is that we have to figure out whose completion to invoke - for example, `du`s. This means we have to expand and parse the command line again, after substituting in the wrapped command. This is doable, it's just a bit of work.

**Comment by inflation** [comment]
Hi, I'm facing the same issue here. For a single `abbr`, it is too long for expansion. And when I have lots of these kind of aliases, I don't have the time to write completion for each of them.

```fish
abbr gwip 'git add -A; git rm (git ls-files --deleted) 2> /dev/null; git commit --no-verify -m "--wip-- [skip ci]"'
```

**Comment by krader1961** [comment]
Personally I recommend using an abbreviation for simple situations like the one in the original problem statement. This is the simplest solution to the problem and works well for me. For example, I have these git related abbreviations:

```
abbr gc 'git commit'
abbr gca 'git commit --amend'
abbr gco 'git checkout'
abbr gcm 'git checkout master'
abbr gd 'git diff'
abbr gs 'git status'
abbr gsb 'git show-branch'
abbr gam 'hub am -3'
abbr gb 'git branch'
```

I can type any of them followed by a space and the subsequent completions work as expected. It is not at all clear how completions should work if the example by @inflation were instead an alias/function.

**Comment by simnalamburt** [comment]
I had created super fast [git abbreviation plugin][cgitc] by [hacking the `abbr` function][hack]. @xnuk found that it becomes much more faster if you update `fish_user_abbreviations` variable directly. 

###### Reference
- [cgitc]
- [cgitc@`05983c7`, init.fish#L40][hack]

[cgitc]: https://github.com/simnalamburt/cgitc
[hack]: https://github.com/simnalamburt/cgitc/blob/05983c7f34/init.fish#L40

**Comment by jacobsa** [comment]
FWIW, I also would really like support for the "git checkout" case, except for a command that doesn't support aliases itself as git does. Abbreviations work, but are suboptimal because the expanded form is really verbose and ugly.

**Comment by despairblue** [comment]
As the person who opened this issue, I have to say that abbreviations seem to me better usability-wise. I've been using them since shortly after opening this issue and they have two main advantages:

1. **I don't forgot what I'm abbreviating.** When I'm on another machine I still know how the command works since abbreviations are more like autocomplete: I know what I want, but I only type it partially.

2. **People see what they execute before they do it.** When someone is using my machine and we share the same alias and they use it, no harm done. But if for some reason I alias `sudo rm -rf /` to `gco` we have a problem. When they know that I use fish they start to look for what it expands to. That's something that highly confuses me when using someone else's bash or zsh setup (typing an alias and not seeing it expanded).

**Comment by krader1961** [comment]
@simnalamburt, Note that fish 3.0.0 will have much faster abbreviations. See issue #4048. That change, however, means that the `fish_user_abbreviations` no longer has any meaning.

**Comment by timothywlewis** [comment]
I have an `git` alias `dsf` (for https://github.com/so-fancy/diff-so-fancy) that should autocomplete the same as `git diff`. Here's the relevant piece of my `~/.gitconfig`:

```
[alias]
	dsf = "!f() { [ -z \"$GIT_PREFIX\" ] || cd \"$GIT_PREFIX\" && git diff --color \"$@\" | diff-so-fancy  | less --tabs=4 -RFX; }; f"
```

My current approach is to clone the '### diff' section of https://github.com/fish-shell/fish-shell/blob/master/share/completions/git.fish into my own `~/.config/fish/completions/git.fish`:

```
# source shared git completions for __fish_git_needs_command and __fish_git_using_command
source $__fish_datadir/completions/git.fish

### dsf is an alias that wraps diff -- https://github.com/so-fancy/diff-so-fancy
###  cloned from 'diff' section in /usr/local/Cellar/fish/2.6.0/share/fish/completions/git.fish
complete -c git -n '__fish_git_needs_command' -a dsf -d 'Show changes between commits, commit and working tree, etc'
complete -c git -n '__fish_git_using_command dsf' -a '(__fish_git_ranges)' -d 'Branch'
complete -c git -n '__fish_git_using_command dsf' -l cached -d 'Show diff of changes in the index'
complete -c git -n '__fish_git_using_command dsf' -l no-index -d 'Compare two paths on the filesystem'
# TODO options
```

Is this the cleanest way to do it? Or is there a cleaner way to inherit all the completions defined for `git diff`?

**Comment by ridiculousfish** [comment]
fish supports wrapping individual commands, but not an entire command line (i.e. including subcommands). Some day we'll support that = see #1976.

**Comment by zanchey** [comment]
That's probably as good as it gets.

**Comment by ridiculousfish** [comment]
This should be working much better in bb7b6491326b19ed9d79e7aaf555b93d98e3e406. Please give it a try and let me know if any aliases aren't behaving properly.

**Comment by Siilwyn** [comment]
@ridiculousfish not sure if I should create a new issue or reply here... this doesn't seem to work with npm subcommands.

For example `alias npr='npm run'` and typing `npr ` gives me completion for all `npm` commands instead of what just the npm scripts that I do get if I type `npm run`.

**Comment by faho** [comment]
@Siilwyn: Are you sure you are running fish 3.0.0?

**Comment by Siilwyn** [comment]
Sorry about that, I figured since this issue was closed a while ago I would be running a version that includes that fix but I'm still on 2.7.

**Comment by timothywlewis** [comment]
So now in fish 3.0.0 can I get `git diff` completions for my git alias? I'd like to get `git dsf <tab>` to have the identical completions as `git diff <tab>`.

```
# this didn't work
complete -c 'git dsf' -w 'git diff'
# and neither did this
complete -c git -n '__fish_seen_subcommand_from dsf' -w 'git diff'
```

The closest I can get to it is a function that wraps it:

```
function gdsf -w 'git diff'
    git dsf $argv;
end
```

Thanks!

> I have an `git` alias `dsf` (for https://github.com/so-fancy/diff-so-fancy) that should autocomplete the same as `git diff`. Here's the relevant piece of my `~/.gitconfig`:
> 
> ```
> [alias]
> 	dsf = "!f() { [ -z \"$GIT_PREFIX\" ] || cd \"$GIT_PREFIX\" && git diff --color \"$@\" | diff-so-fancy  | less --tabs=4 -RFX; }; f"
> ```
> My current approach is to clone the '### diff' section of https://github.com/fish-shell/fish-shell/blob/master/share/completions/git.fish into my own `~/.config/fish/completions/git.fish`:
> 
> ```
> # source shared git completions for __fish_git_needs_command and __fish_git_using_command
> source $__fish_datadir/completions/git.fish
> 
> ### dsf is an alias that wraps diff -- https://github.com/so-fancy/diff-so-fancy
> ###  cloned from 'diff' section in /usr/local/Cellar/fish/2.6.0/share/fish/completions/git.fish
> complete -c git -n '__fish_git_needs_command' -a dsf -d 'Show changes between commits, commit and working tree, etc'
> complete -c git -n '__fish_git_using_command dsf' -a '(__fish_git_ranges)' -d 'Branch'
> complete -c git -n '__fish_git_using_command dsf' -l cached -d 'Show diff of changes in the index'
> complete -c git -n '__fish_git_using_command dsf' -l no-index -d 'Compare two paths on the filesystem'
> # TODO options
> ```
> Is this the cleanest way to do it? Or is there a cleaner way to inherit all the completions defined for `git diff`?

**Comment by Siilwyn** [comment]
@timothywlewis not sure what you're trying to achieve as `git dsf ` is just as long to type as `git diff`? Either way in 3.0 you can create an alias for `git diff` and hit tab to indeed get auto-completion.

**Comment by zanchey** [comment]
fish will complete aliases that only use internal git functions, but not those that start shell pipelines; your current method is probably the cleanest for now.
